chore(ruff): Enable and fix all ruff rules that we can

bellini666 opened the pull request:
I went through all ruff rules we were ignoring and enabled all that made sense.

Some have legit use cases in some places, so I changed to an inline "# noqa <rule>" instead of having to disable those for the whole file. Some still made sense to be ignored on tests.

## Summary by Sourcery

Enable more Ruff rules and fix existing violations.

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @_bellini666 for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```
<!-- action-check: release-file -->

bellini666 commented on strawberry/types/auto.py:
```diff
@@ -34,7 +34,7 @@ def __call__(cls, *args: str, **kwargs: Any) -> Any:
         return cls._instance
 
     def __instancecheck__(
-        self,
+        cls,
         instance: Union[StrawberryAuto, StrawberryAnnotation, StrawberryType, type],
     ) -> bool:
```
`__instancecheck__` is a special case of a classmethod

patrick91 commented on strawberry/aiohttp/test/client.py:
```diff
@@ -46,7 +46,8 @@ async def query(
 
         if assert_no_errors:
-            assert resp.status == 200
+            ok_response = 200
+            assert resp.status == ok_response
```
I don't like this :P

bellini666 replied:
This is the "don't use magic numbers" rule. TBH I'm not a huge fan as well, let's ignore it?

patrick91 commented on strawberry/cli/commands/codegen.py:
```diff
@@ -39,25 +39,18 @@ def _import_plugin(plugin: str) -> Optional[type[QueryCodegenPlugin]]:
         }
 
-        if "__all__" in module.__dict__:
-            symbols = {
-                name: symbol
-                for name, symbol in symbols.items()
-                if name in module.__dict__["__all__"]
-            }
-
-        for obj in symbols.values():
-            if _is_codegen_plugin(obj):
-                return obj
-
-    return None
+    return next((obj for obj in symbols.values() if _is_codegen_plugin(obj)), None)
```
same here, it feels harder to understand now 😊

patrick91 commented on strawberry/django/views.py:
```diff
@@ -266,7 +265,7 @@ class AsyncGraphQLView(
     request_adapter_class = AsyncDjangoHTTPRequestAdapter
 
     @classonlymethod  # pyright: ignore[reportIncompatibleMethodOverride]
-    def as_view(cls, **initkwargs: Any) -> Callable[..., HttpResponse]:
+    def as_view(self, **initkwargs: Any) -> Callable[..., HttpResponse]:
```
this should still be `cls` :)

bellini666 replied:
oh, automatic ruff change fail

patrick91 commented on strawberry/relay/utils.py:
```diff
@@ -33,8 +33,9 @@ def from_base64(value: str) -> tuple[str, str]:
 
-    if len(res) != 2:
-        raise ValueError(f"{res} expected to contain only 2 items")
+    expected_len = 2
+    if len(res) != expected_len:
+        raise ValueError(f"{res} expected to contain only {expected_len} items")
```
I don't link this change too :D

erikwrede commented on strawberry/cli/commands/server.py:
```diff
@@ -54,13 +54,13 @@ def server(
             r"pip install 'strawberry-graphql\[debug-server]'"
         )
-        raise typer.Exit(1)
+        raise typer.Exit(1) from e
```
I disagree about raising from e here, cluttering the error output. We are catching this error as expected and print a text.

patrick91 replied:
I'm not even sure it is printed anywhere

bellini666 replied:
Hrm, I can ignore the ruff rule for this line in specific

erikwrede commented on strawberry/cli/utils/__init__.py:
```diff
@@ -16,7 +16,7 @@ def load_schema(schema: str, app_dir: str) -> Schema:
 
         rich.print(f"[red]Error: {message}")
-        raise typer.Exit(2)
+        raise typer.Exit(2) from exc
```
same case as above with exception raising in cli

erikwrede commented on strawberry/chalice/views.py:
```diff
@@ -118,7 +118,7 @@ def create_response(
     ) -> Response:
         status_code = 200
 
-        if sub_response.status_code != 200:
+        if sub_response.status_code != status_code:
             status_code = sub_response.status_code
```
Looking at this again, why don't we set the status to sub_response.status code immediately? seems like an unnecessary statement this way

bellini666 replied:
This was mostly to fix the "do not compare with magic numbers", I haven't checked the actual code. Will see if I can refactor it :)

bellini666 commented on .github/release-check-action/release.py:
```diff
@@ -35,10 +35,10 @@ def get_release_info(file_path: Path) -> ReleaseInfo:
         change_type_key = match.group(1)
         change_type = ChangeType[change_type_key.upper()]
-        changelog = "".join([line for line in f.readlines()]).strip()
+        changelog = "".join(list(f.readlines())).strip()
```
hahaha ruff did this change, agreed!

erikwrede commented on strawberry/extensions/context.py:
```diff
@@ -128,41 +128,40 @@ async def iterator() -> AsyncIterator:
         if iscoroutinefunction(func):
+            is_async = True
 
             @contextlib.asynccontextmanager
             async def iterator() -> AsyncIterator[None]:
                 await func(extension)
                 yield
 
-            return WrappedHook(extension=extension, hook=iterator, is_async=True)
         else:
+            is_async = False
 
             @contextlib.contextmanager
             def iterator() -> Iterator[None]:
                 func(extension)
                 yield
 
-            return WrappedHook(extension=extension, hook=iterator, is_async=False)
+        return WrappedHook(extension=extension, hook=iterator, is_async=is_async)
```
This is more redundant and harder to read than before

erikwrede commented on strawberry/schema/base.py:
```diff
@@ -88,7 +87,6 @@ def get_type_by_name(
         raise NotImplementedError
 
     @abstractmethod
-    @lru_cache
```
just flagging this to come back & check the children later

erikwrede commented on strawberry/schema/schema.py:
```diff
@@ -283,7 +285,7 @@ def _create_execution_context(
         )
 
-    @lru_cache
+    @lru_cache  # noqa: B019
```
https://docs.astral.sh/ruff/rules/cached-instance-method/

very sceptical about enabling the rule here, as many instances here have potentially infinite lifetime and we _need_ caching on them. My concern is it may discourage PR submitters from caching where it makes sense

bellini666 replied:
Good point! Going to disable this one globally

erikwrede commented on strawberry/schema/types/base_scalars.py:
```diff
@@ -15,16 +15,18 @@ def inner(value: str) -> object:
-            raise GraphQLError(f'Value cannot represent a {type_}: "{value}". {e}')
+            raise GraphQLError(
+                f'Value cannot represent a {type_}: "{value}". {e}'
+            ) from e
```
I feel like this goes against our opinionation of how we wrap GraphQL core. It makes more sense to me to keep hiding this, as we want to be the ones to decide the error output. Of course at the added cost of debuggability.

erikwrede commented on strawberry/schema/types/base_scalars.py:
```diff
@@ -15,16 +15,18 @@ def inner(value: str) -> object:
-    except decimal.DecimalException:
-        raise GraphQLError(f'Value cannot represent a Decimal: "{value}".')
+    except decimal.DecimalException as e:
+        raise GraphQLError(f'Value cannot represent a Decimal: "{value}".') from e
```
Also, I don't think GraphQL Errors should be raised from anything. The spec errors are good in their closed form, I'd prefer them to not be related to the python errors.

erikwrede commented on tests/experimental/pydantic/test_conversion.py:
```diff
@@ -1257,10 +1257,10 @@ class User:
 
 
-SI = TypeVar("SI", covariant=True)  # pragma: no mutate
+SI_co = TypeVar("SI_co", covariant=True)  # pragma: no mutate
```
Which rule was this?

erikwrede replied:
https://docs.astral.sh/ruff/rules/type-bivariance/

not sure about this one

bellini666 replied:
Hrm, I don't have any strong preferences here. I can revert and ignore this if you prefer :)

erikwrede replied:
I think not using this rule leads to a better code style, since our GraphQL generics are covariant by default. Adding the suffix disturbs the flow a bit

erikwrede commented on tests/fields/test_field_defaults.py:
```diff
@@ -55,10 +55,10 @@ class Query:
 
 def test_field_with_separate_resolver_default():
-    with pytest.raises(FieldWithResolverAndDefaultValueError):
+    def gun_resolver() -> str:
+        return "revolver"
```
this is such a twist on the eye resolver revolver xD

erikwrede replied:
we should probably change this to some strawberry themed test :D

erikwrede reviewed (commented):
reviewed everything but tests, some comments are very opinionated

bellini666 commented:
@patrick91 @erikwrede thank you for the careful review :) (the PR is large lol)

I reverted the changes for the rules that we don't want in the codebase and added them to the ignore list.

Let me know if there's anything else you want me to revert here

erikwrede reviewed (approved):
lgtm now :) @patrick91 what do you think?
